File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys

input = sys.stdin.readline

n, k = map(int, input().split())
w, v = [0] * 101, [0] * 101
dp = [[0 for _ in range(100_001)] for _ in range(101)]

goods = []
for i in range(1, n + 1):
    w[i], v[i] = map(int, input().split())

for i in range(1, n + 1):
    for j in range(1, k + 1):
        if j < w[i] : 
            dp[i][j] = dp[i - 1][j]
        else :
            dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - w[i]] + v[i])

print(dp[n][k])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.rs
use std::io::{stdin, stdout, Read, Write, BufReader, BufWriter};
use std::cmp::*;

fn main() {
    let mut reader = BufReader::new(stdin());
    
    let mut input = String::new();
    reader.read_to_string(&mut input).unwrap();
    let mut input = input.split_ascii_whitespace().flat_map(str::parse::<usize>);

    let (n, k) = (input.next().unwrap(), input.next().unwrap());
    let (mut w, mut v) = (vec![0; 101], vec![0; 101]);
    let mut dp = vec![vec![0; 100_001]; 101];

    for i in 1..=n {
        w[i] = input.next().unwrap();
        v[i] = input.next().unwrap();
    }

    for i in 1..=n {
        for j in 1..=k {
            if j < w[i] {
                dp[i][j] = dp[i - 1][j];
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - w[i]] + v[i]);
            }
        }
    }

    println!("{}", dp[n][k]);
}
